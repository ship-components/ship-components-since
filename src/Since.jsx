// External
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ms from 'ms';

/**
 * Displays a counter since a timestamp
 * @type    {React}
 */
export default class Since extends Component {

  /**
   * Setup
   * @param    {Object}    props
   */
  constructor(props) {
    super(props);
    this.loop = this.loop.bind(this);
  }

  /**
   * Start
   */
  componentDidMount() {
    this.loop();
  }

  /**
   * Performance check
   * @param    {Object}    nextProps
   * @return   {Boolean}
   */
  shouldComponentUpdate(nextProps) {
    return nextProps.timestamp !== this.props.timestamp ||
           nextProps.className !== this.props.className;
  }

  /**
   * Cleanup
   */
  componentWillUnmount() {
    clearTimeout(this.updateId);
  }

  /**
   * Causes a forcedUpdate so view actually updates
   */
  loop() {
    // The meat and potatoes
    this.forceUpdate();

    // Loop
    this.updateId = setTimeout(this.loop, this.getInterval());
  }

  /**
   * Determine how frequently to update the component
   * @return    {[type]}
   */
  getInterval() {
    const diff = Math.abs(Date.now() - this.props.timestamp);

    // If the timestamp just happened update more frequently
    if (diff < 1000) {
      return 300;
    } else if (diff < 60000) {
      return 1000;
    }

    return 10000;
  }

  /**
   * Render
   * @return    {React}
   */
  render() {
    const {
      timestamp,
      className,
      showPrefix,
      showSuffix
    } = this.props;

    // Get the ms since the timestamp
    const diff = Math.abs(Date.now() - timestamp);

    // Determine if it is after or before now
    const inFuture = Date.now() < timestamp;

    return (
      <this.props.tag
        className={classNames(className)}
      >
        {inFuture && showPrefix ? 'in ' : ''}
        {diff < 1000 ? 0 : ms(diff)}
        {!inFuture && showSuffix ? ' ago' : ''}
      </this.props.tag>
    );
  }
}

/**
 * Defauts
 * @type    {Object}
 */
Since.defaultProps = {
  showPrefix: true,
  showSuffix: false,
  className: undefined,
  tag: 'span'
};

/**
 * Type Checks
 * @type    {Object}
 */
Since.propTypes = {
  showPrefix: PropTypes.bool,
  showSuffix: PropTypes.bool,
  timestamp: PropTypes.number.isRequired,
  className: PropTypes.string,
  tag: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node
  ])
};
