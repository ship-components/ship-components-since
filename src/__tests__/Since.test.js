/* eslint-disable react/no-multi-comp */

import React from 'react';
import { mount } from 'enzyme';
import ms from 'ms';
import renderer from 'react-test-renderer';

import Since from '../Since';

const ExpectedValues = {
  '1s' : ms('1s'),
  '30s' : ms('30s'),
  '1m' : ms('1m'),
  '1h' : ms('1h'),
  'in 30m' : ms('30m') * -1
};

Object.keys(ExpectedValues).forEach((fuzzyTime) => {
  test(`it should render ${fuzzyTime} when the difference from now and the timestamp is ${ExpectedValues[fuzzyTime]}ms`, () => {
    const tree = renderer.create(
      <Since
        timestamp={Date.now() - ExpectedValues[fuzzyTime]}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});


test('it should add "ago" if the showSuffix prop is supplied', () => {
  const wrapper = mount(
    <Since
      showSuffix
      timestamp={Date.now() - 10000}
    />
  );
  expect(wrapper.html().includes('ago')).toBeTruthy();
});

test('it should remove "in" if the showPrefix prop is false', () => {
  const wrapper = mount(
    <Since
      showPrefix={false}
      timestamp={Date.now() - 10000}
    />
  );
  expect(wrapper.html().includes('in')).toBeFalsy();
});

test('it should change the update frequency depending on how close to the timestamp Date.now() is', () => {
  const wrapper = mount(
    <Since
      timestamp={Date.now()}
    />
  );
  // Sub 1s
  expect(wrapper.get(0).getInterval()).toBe(300);

  // Sub 1m
  wrapper.setProps({
    timestamp: Date.now() + 2000
  });
  expect(wrapper.get(0).getInterval()).toBe(1000);

  // Greater than 1m
  wrapper.setProps({
    timestamp: Date.now() + 20000000
  });
  expect(wrapper.get(0).getInterval()).toBe(10000);
});

test('it should support custom tag types', () => {
  const wrapper = mount(
    <Since
      tag='div'
      timestamp={Date.now()}
    />
  );
  expect(wrapper.find('span').length).toBe(0);
  expect(wrapper.find('div').length).toBe(1);
});

test('it should clear the timeout when it unmounts', () => {
  const wrapper = mount(
    <Since
      tag='div'
      timestamp={Date.now()}
    />
  );
  const _clearTimeout = window.clearTimeout;
  window.clearTimeout = jest.fn();

  expect(typeof wrapper.get(0).updateId).toBe('number');
  wrapper.unmount();
  expect(window.clearTimeout.mock.calls.length).toBe(1);

  window.clearTimeout = _clearTimeout;
});

test('it should update when the className or timestamp changes', () => {
  const props = {
    timestamp: Date.now(),
    className: 'test'
  };
  const shouldComponentUpdate = Since.prototype.shouldComponentUpdate.bind({ props });
  expect(shouldComponentUpdate(Object.assign({}, props, {
    className: 'changed'
  }))).toBe(true);
  expect(shouldComponentUpdate(Object.assign({}, props, {
    timestamp: Date.now() + 1000
  }))).toBe(true);
  expect(shouldComponentUpdate(props)).toBe(false);
});
