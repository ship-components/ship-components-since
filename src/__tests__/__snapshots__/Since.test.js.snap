// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render 1h when the difference from now and the timestamp is 3600000ms 1`] = `
<span
  className=""
>
  
  1h
  
</span>
`;

exports[`it should render 1m when the difference from now and the timestamp is 60000ms 1`] = `
<span
  className=""
>
  
  1m
  
</span>
`;

exports[`it should render 1s when the difference from now and the timestamp is 1000ms 1`] = `
<span
  className=""
>
  
  1s
  
</span>
`;

exports[`it should render 30s when the difference from now and the timestamp is 30000ms 1`] = `
<span
  className=""
>
  
  30s
  
</span>
`;

exports[`it should render in 30m when the difference from now and the timestamp is -1800000ms 1`] = `
<span
  className=""
>
  in 
  30m
  
</span>
`;
